djstripe/models.py

```diff
@@ -2872,32 +2792,133 @@ def is_expired(self):
+        from .utils import fix_django_headers
 
-    def __str__(self):
-        return smart_text("<{message}, pk={pk}, Event={event}>".format(
-            message=self.message,
-            pk=self.pk,
-            event=self.event
-        ))
+        headers = fix_django_headers(request.META)
+        assert headers
+        try:
+            body = request.body.decode(request.encoding or "utf-8")
+        except Exception:
+            body = "(error decoding body)"
+
+        ip = request.META["REMOTE_ADDR"]
+        obj = cls.objects.create(headers=headers, body=body, remote_ip=ip)
+
+        try:
+            obj.valid = obj.validate()
+            if obj.valid:
+                if djstripe_settings.WEBHOOK_EVENT_CALLBACK:
+                    # If WEBHOOK_EVENT_CALLBACK, pass it for processing
+                    djstripe_settings.WEBHOOK_EVENT_CALLBACK(obj)
+                else:
+                    # Process the item (do not save it, it'll get saved below)
+                    obj.process(save=False)
+        except Exception as e:
+            max_length = WebhookEventTrigger._meta.get_field("exception").max_length
+            obj.exception = str(e)[:max_length]
+            obj.traceback = format_exc()
+
+            # Send the exception as the webhook_processing_error signal
+            webhook_processing_error.send(
+                sender=WebhookEventTrigger, exception=e, data=getattr(e, "http_body", "")
+            )
+        finally:
+            obj.save()
+
+        return obj
+
+    @cached_property
+    def json_body(self):
+        import json
+
+        try:
+            return json.loads(self.body)
+        except ValueError:
+            return {}
+
+    @property
+    def is_test_event(self):
+        return self.json_body.get("id") == webhooks.TEST_EVENT_ID
+
+    def validate(self, api_key=None):
+        """
+        The original contents of the Event message must be confirmed by
+        refetching it and comparing the fetched data with the original data.
+
+        This function makes an API call to Stripe to redownload the Event data
+        and returns whether or not it matches the WebhookEventTrigger data.
+        """
+
+        local_data = self.json_body
+        if "id" not in local_data or "livemode" not in local_data:
+            return False
+
+        if self.is_test_event:
+            logger.info("Test webhook received: {}".format(local_data))
+            return False
+
+        livemode = local_data["livemode"]
+        api_key = api_key or djstripe_settings.get_default_api_key(livemode)
+
+        # Retrieve the event using the api_version specified in itself
+        with stripe_temporary_api_version(local_data["api_version"], validate=False):
+            remote_data = Event.stripe_class.retrieve(id=local_data["id"], api_key=api_key)
+
+        return local_data["data"] == remote_data["data"]
+
+    def process(self, save=True):
+        self.event = Event.process(self.json_body)
```

Comment:
It'd be nice to not hardcode `Event` here directly for those crazy people that want to extend the functionality without re-implementing anything else. Perhaps we could make the model for event configurable? Anyway, not directly related to the code changes here since the functionality wasn't there before either.

Reply:
I see where you're going with this but I don't think extending the djstripe Stripe classes is a good idea. If you wanted to extend the functionality, you'd subclass the WebhookEventTrigger class imo.

Reply:
I'm leaning with @jleclanche on this one. I suppose we should add ``event_class`` as a class variable. to make the subclassing less painful for those choosing to do so.